docs/architecture/adr-011-light-client-crates-extraction.md

```diff
@@ -0,0 +1,375 @@
+    fn client_def(&self) -> Box<dyn ClientDef>;
+
+    /* ... */
+}
+```
+
+Now, we can use the `ClientState` instance to get the `ClientDef` implementation ->
+
+```rust
+pub fn process(
+    ctx: &dyn ClientReader,
+    /* ... */
+) -> HandlerResult<ClientResult, Error> {
+    /* ... */
+
+    let client_state = ctx.client_state(&client_id)?;
+
+    let client_def = client_state.client_def();
+
+    /* ... */
+}
+```
+
+### Domain types containing light client specific types
+
+Raw types such as `MessageCreateClient` contain light client `ClientState` and `ConsensusState` serialized
+as `google::protobuf::Any`.
+
+```rust
+pub struct MsgCreateClient {
+    /// light client state
+    #[prost(message, optional, tag = "1")]
+    pub client_state: ::core::option::Option<super::super::super::super::google::protobuf::Any>,
+    /// consensus state associated with the client that corresponds to a given
+    /// height.
+    #[prost(message, optional, tag = "2")]
+    pub consensus_state: ::core::option::Option<super::super::super::super::google::protobuf::Any>,
+    /* ... */
+}
+```
+
+Ideally, the domain type would contain validated `ClientState` and `ConsensusState` types but this is not possible
+anymore because the `ibc` crate cannot depend on the light client crates. It is therefore proposed that domain types
+continue to use the `Any` type for such fields and defer the validation to the handlers.
+
+### Light client specific code
+
+A sizeable amount of light client specific code exists in the `ibc` crate exists today that is only used by `hermes`.
+For e.g. helper functions that extract `IbcEvent`s from `tendermint::abci::Event`s (i.e. `from_tx_response_event()`).
+
+There are cases where core types depend on light client types, for e.g. `ics02_client::Error`
+uses `ics07_tendermint::Error`s in some of its variants.
+
+Additionally, some of that code might be helpful for host implementations, for e.g. event conversions
+to/from `tendermint::abci::Event` are required for `deliverTx` implementation.
+
+It is recommended that most client specific code is moved either to the `ibc-relayer` crate or the light client crates
+and issues of this kind be dealt with on a case-by-case basis.
+
+### Light client registry
+
+With the proposals in this ADR, the `ibc` crate would be light client agnostic, however, the host implementation must
+still be aware of all light clients that it wishes to support. For e.g., after an upgrade a blockchain must be able to
+deserialize types from the persistent store.
+Furthermore, protobuf has emerged as the canonical serialization scheme for IBC, and IBC's message definitions usually
+serialize light client types using the `google::protobuf::Any` type where the `type_url` is accepted to uniquely
+represent specific light client types, although this has not been standardized yet.
+It is proposed to standardize this and provide a `ibc-client-registry` crate which standardizes this and collects all
```

Comment:
A cool crate that might be helpful in implementing such a registry -> https://github.com/dtolnay/inventory. (Thanks to @romac for suggesting it).